### ISSUE TITLE
Add simple deny-listing capability to distributed queries

### ISSUE THREAD

**Issue by lucasmrod** [osquery/osquery#7658]
Hi folks. I'm Lucas from [Fleet](https://github.com/fleetdm).

I'll first introduce the issue we are seeing with distributed queries, and later propose how osquery could help with such scenario.

## What

Following is the expected/happy-path scenario for osquery interacting with the Fleet server:

1. Osquery requests queries via the `distributed/read` request.
2. Fleet server returns queries to execute.
3. Osquery executes the queries sequentially.
4. Osquery sends the results of all the queries via the `distributed/write` request.

That said, we've seen the following scenario on some hosts/deployments:

1. Osquery requests queries via the `distributed/read` request.
2. Fleet server returns queries to execute.
3. Osquery start executing the queries sequentially.
4. Some query exceeds a performance threshold and the worker process is killed by the watcher process (e.g. memory or CPU intensive query).
5. Osquery worker restarts.

(Osquery would run steps 1 to 5 over and over, and the Fleet server has no idea what happened or which query is the problematic one.)

## Proposal

A simple proposal to fix this would be to add "denylisting" capability to "distributed queries", here's some pseudo-code:
```
Distributed::runQueries() {
  for query in queries {
    
    if query is already marked as running (in RocksDB)
      set its result as failed with "denylisted" message
      add result to results
      continue with next query in the loop
     
    mark query as running (in RocksDB)
    run query
    mark query as not running (in RocksDB)
    add result to results
  }
  send results
}
```
The queries would be stored as running (i.e. denylisted) for a configured time (and a cleanup method would remove expired entries).

## Why

By adding deny-listing capabilities to distributed queries, we could:
1. Avoid continuous crashes of osquery worker processes.
2. Provide more information to a server about which query is the problematic one.

**Comment by zwass** [comment]
One thought I have is:

Let's say queries A, B, and C are currently "active" on the server (meaning the server will send them when the host requests the set of queries to run). B and C are both "expensive" (meaning the watchdog will kill their execution).

In this scenario, do we end up in the situation where query A is executed 3 times before osquery replies to the server with results for A and failures for B and C?

Should we try to mitigate that, or do you think I'm overoptimizing here?

**Comment by lucasmrod** [comment]
> In this scenario, do we end up in the situation where query A is executed 3 times before osquery replies to the server with results for A and failures for B and C?

Ah good catch, yes, query `A` would be executed three times in the host.

> Should we try to mitigate that, or do you think I'm overoptimizing here?

A way to mitigate this would be to store to-be-sent distributed results in RocksDB instead of in memory. So, before running the query, it checks if there are unsubmitted results for it, if so, use those instead of running the query. But would make this issue+PR a bit more complex, I can create a separate issue for it if that makes sense.

**Comment by Smjert** [comment]
Hello @lucasmrod, I think there's another thorny issue here, which is already present with the current system of denylisting: the watchdog kills osquery when the whole worker process is using, for instance, CPU over the limit for a certain amount of time.

Coupled with the fact that scheduled queries, events and distributed queries will run in parallel, this means that if you have events enabled and gathering events takes 80% of the CPU limit, then comes a scheduled query which take the rest, the scheduled query will get denylisted, even though it wasn't the real cause of the high CPU usage.

And to clarify, what I mean with those percentages is not that it consumes 80% of the CPU usage limit, so if the CPU usage limit is 40% of one core, then we calculate the 80% of that.
It's more complex because the watchdog considers **sustained** cpu usage, with a sample every 3 seconds and the limit is X CPU usage for 12 seconds.
So going back to my example, for the events that would mean for instance using more than 40% of CPU for around 9.6 seconds (80% of 12), and the rest of the time, the same CPU usage but from the scheduled query.

The same can happen for a distributed query which goes beyond the CPU usage threshold for 80% of the time threshold. Then comes again a scheduled query which gets over the CPU usage threshold for 20% of the time threshold. The scheduled query here would get denylisted.

The change proposed here would not change anything in those cases, but would then create a new case where events or scheduled queries could cause distributed queries to push osquery "over the edge" of the limit, and be denylisted incorrectly.

Sorry if I'm not proposing alternatives but it's indeed difficult, because with what we have now we don't have a good way to separate the overhead of each operation and so understand what really caused the watchdog to kill osquery (without actively looking).

The ideal would be to have somehow a way to detect the contribution from each operation into causing osquery to get over the limit, and then use some strategy to either reorder them in a way that they don't overlap, or if there's an operation that is clearly causing a lot of overhead (like I don't know more than 50%), then one temporarily block that one.

I don't have specific suggestions on how to do it, due to again the difficulty of separating the overhead metrics, but I wanted to know what were your thought about the above issues.

**Comment by lucasmrod** [comment]
Thanks for the context! It really helps understand the big picture.

> The change proposed here would not change anything in those cases, but would then create a new case where events or scheduled queries could cause distributed queries to push osquery "over the edge" of the limit, and be denylisted incorrectly.

So we have two scenarios:
A. A distributed query is indeed exceeding a performance threshold and causes the watcher to kill the worker.
B. Performance threshold is exceeded during a distributed query execution, but it's caused by other query/work running simultaneously (e.g. a scheduled query), and causes the watcher to kill the worker. (False positive.)

AFAICS,
- The proposed change helps with case (A).
- The proposed change adds an "expiration time" to the deny-listed distributed queries to eventually recover from case (B).

> I don't have specific suggestions on how to do it, due to again the difficulty of separating the overhead metrics, but I wanted to know what were your thoughts about the above issues.

The best isolation I can think of would be to have two worker processes:
- A worker process that runs the "distributed thread" (uses its own separate RocksDB for its own "distributed" functionality). It could also open the main RocksDB database in read mode only, mostly to read configs/etc.
- Another worker process that the rest, scheduled queries thread, eventing, etc.

And have a single watcher monitor them, or have each worker have its own watcher.
I might be missing some dependencies or limitations of doing this.

--

Another (maybe simpler?) idea would be to not allow running scheduled queries at the same time as distributed queries, proper prioritization should be discussed, but this would also allow better detection of which query is exceeding performance thresholds.

**Comment by Smjert** [comment]
> The best isolation I can think of would be to have two worker processes:

One of the problems with splitting it in multiple processes is that you would need to have the logic of the tables to exists in both processes. This means that with the tables that cache data in memory for the duration of the process, you will have doubled the memory cost.

We also have services that do things at the worker startup to fill up those caches (the recently committed users and groups optimization for Windows DCs); that work would be duplicated and cause double the CPU usage to gather user and groups data.
I did thought to put this data on file, but since we already have issues with RocksDB and disk usage with events, I wasn't exactly sure of the result. It would also have been a step towards actually being a database that stores data on file.

>Another (maybe simpler?) idea would be to not allow running scheduled queries at the same time as distributed queries, proper prioritization should be discussed, but this would also allow better detection of which query is exceeding performance thresholds.

The concern here though is having scheduled queries or distributed queries which maybe do not use much resources but are long running. Especially for distributed queries, these could delay the execution of scheduled queries by a lot.

Anyway I guess that even though what I described in the previous can happen, if the user is informed, they can check which queries where running in the period around the time where the denylisting happened, and then look closer to those queries.
The data provided by osquery wouldn't help much (for instance osquery_schedule), because that still uses the data from the whole process.

**Comment by directionless** [comment]
We talked about this in office hours today (2022-07-05) and the general sentiment is that we should do the simple thing outlined at the beginning.

Yes, it might incorrectly blame a scheduled query for a distributed one (or vice versa) but we're not sure how likely that is, and more complex things aren't clearly correct either.

**Comment by lucasmrod** [comment]
> Er, I mean... How does the TLS server know? How is this connected to the status?

Ah, see the following line in the proposed [diff](https://github.com/osquery/osquery/compare/master...lucasmrod:add-denylist-mechanism-to-dist-queries?expand=1). Let me know if that's what you mean.

**Comment by lucasmrod** [comment]
#7675 has been merged and released in [5.5.1](https://github.com/osquery/osquery/releases/tag/5.5.1).
